cyberark/conjur
# Use Authenticators to implement both /login and /authenticate

#### What does this PR do?

This PR updates authentication to implement both `/login` and `/authenticate` with authenticators. It also updates the LDAP authenticator to support `/authenticate` with LDAP password (the original, now deprecated, behavior) or with the Conjur API key return by LDAP authentication `/login`

#### Any background context you want to provide?

There are guiding principals for these Conjur authentication changes:
1) People users should `/login` with a password and services should `/authenticate` with a key or token.
2) `/login` should return with a token that is accepted by `/authenticate` for the same authenticator
3) Authenticators generally MUST implement `/authenticate`, but may implement `/login`
4) To be used by the Conjur UI, an authenticator MUST implement `/login`

#### What ticket does this PR close?

Closes https://github.com/conjurinc/conjur-ui/issues/204

#### Where should the reviewer start?

Reading the updated `AUTHENTICATORS.md` readme then walking through the commits is probably the best way to understand the intention behind this PR and the steps I took to implement it. The commits have been cleaned up into as discreet of steps as I could make them.

#### How should this be manually tested?
These changes have automated tests in the CI pipeline.

If you wanted to manually test this, the easiest way might be to run the Conjur UI from this branch https://github.com/conjurinc/conjur-ui/compare/mlee-authn-ldap. This branch expects to be able to login using LDAP, and the user `alice` with password `alice` should work.

#### Link to build in Jenkins (if appropriate)
https://jenkins.conjur.net/job/cyberark--conjur/job/authenticator_backed_login/

#### Has the Version and Changelog been updated?
Yes

#### Questions:
> Does this work have automated integration and unit tests?

Yes

> Can we make a blog post, video, or animated GIF of this?

I'm not sure.

> Has this change been documented (Readme, docs, etc.)?

The developer README is updated

#### TODO:

**Coding**
- [x] Refine design for authenticators to implement `/login` vs `/authenticate` (e.g. method names, fallback mechanisms, etc.)
- [x] Move primary LDAP logic to `/login` rather than `/authenticate`. `/authenticate` should attempt LDAP but fall back to the default authenticate (with API key).
- [x] Resolve relevant code climate issues
- [x] Clean up history to make it easier to review
- [x] Address PR review comments

**Testing**
- [x] LDAP unit tests should use login rather than authenticate to test password and admin behavior
- [x] Add integration tests for LDAP login and authenticating with returned token

**Documentation**
- [x] Update authenticators documentation
- [x] Consider documentation update needs for https://docs.conjur.org/Latest/en/Content/Operations/Services/Authentication-new.htm
